File: app/src/main/java/org/deiverbum/app/data/entity/LHPsalmodyEntity.java
package org.deiverbum.app.data.entity;

import static org.deiverbum.app.utils.Constants.LH_PSALMODY;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */

@Entity(tableName = LH_PSALMODY,
        primaryKeys = {"groupFK", "readingFK", "antiphonFK"},
        foreignKeys =
                {
                        @ForeignKey(
                                entity = LHPsalmodyJoinEntity.class,
                                parentColumns = "groupID",
                                childColumns = "groupFK",
                                onDelete = ForeignKey.CASCADE,
                                onUpdate = ForeignKey.CASCADE),
                        @ForeignKey(
                                entity = PsalmEntity.class,
                                parentColumns = "psalmID",
                                childColumns = "readingFK",
                                onDelete = ForeignKey.CASCADE,
                                onUpdate = ForeignKey.CASCADE),
                        @ForeignKey(
                                entity = LHAntiphonEntity.class,
                                parentColumns = "antiphonID",
                                childColumns = "antiphonFK",
                                onDelete = ForeignKey.CASCADE,
                                onUpdate = ForeignKey.CASCADE),
                        @ForeignKey(
                                entity = LHThemeEntity.class,
                                parentColumns = "themeID",
                                childColumns = "themeFK",
                                onDelete = ForeignKey.SET_NULL,
                                onUpdate = ForeignKey.CASCADE),
                        @ForeignKey(
                                entity = EpigraphEntity.class,
                                parentColumns = "epigraphID",
                                childColumns = "epigraphFK",
                                onDelete = ForeignKey.SET_NULL,
                                onUpdate = ForeignKey.CASCADE)
                }
)
public class LHPsalmodyEntity {

    @NonNull
    @ColumnInfo(name = "groupFK")
    public Integer grupoFK = 0;

    @NonNull
    @ColumnInfo(name = "readingFK", index = true)
    public Integer salmoFK = 0;

    @NonNull
    @ColumnInfo(name = "order")
    public Integer orden = 0;

    @NonNull
    @ColumnInfo(name = "antiphonFK", defaultValue = "0", index = true)
    public Integer antifonaId = 0;

    @ColumnInfo(name = "themeFK", defaultValue = "NULL", index = true)
    public Integer temaFK = 0;

    @ColumnInfo(name = "epigraphFK", defaultValue = "NULL", index = true)
    public Integer epigrafeFK = 0;

    @ColumnInfo(name = "part", defaultValue = "NULL")
    public Integer parte = 0;

    @NonNull
    public Integer getGrupoFK() {
        return grupoFK;
    }

    public void setGrupoFK(@NonNull Integer grupoFK) {
        this.grupoFK = grupoFK;
    }

    @NonNull
    public Integer getSalmoFK() {
        return salmoFK;
    }

    public void setSalmoFK(@NonNull Integer salmoFK) {
        this.salmoFK = salmoFK;
    }

    @NonNull
    public Integer getOrden() {
        return orden;
    }

    public void setOrden(@NonNull Integer orden) {
        this.orden = orden;
    }

    @SuppressWarnings("unused")
    @NonNull
    public Integer getAntifonaId() {
        return antifonaId;
    }

    @SuppressWarnings("unused")
    public void setAntifonaId(@NonNull Integer antifonaId) {
        this.antifonaId = antifonaId;
    }

    public String getParte() {
        return parte == null || parte == 0 ? "" : String.valueOf(parte);
    }


}

File: app/src/main/java/org/deiverbum/app/data/entity/LHPsalmodyEntity.kt
package org.deiverbum.app.data.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.ForeignKey.Companion.CASCADE
import androidx.room.ForeignKey.Companion.SET_NULL
import org.deiverbum.app.utils.Constants

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
@Entity(
    tableName = Constants.LH_PSALMODY,
    primaryKeys = ["groupFK", "readingFK", "antiphonFK"],
    foreignKeys = [ForeignKey(
        entity = LHPsalmodyJoinEntity::class,
        parentColumns = arrayOf("groupID"),
        childColumns = arrayOf("groupFK"),
        onDelete = CASCADE,
        onUpdate = CASCADE
    ), ForeignKey(
        entity = PsalmEntity::class,
        parentColumns = arrayOf("psalmID"),
        childColumns = arrayOf("readingFK"),
        onDelete = CASCADE,
        onUpdate = CASCADE
    ), ForeignKey(
        entity = LHAntiphonEntity::class,
        parentColumns = arrayOf("antiphonID"),
        childColumns = arrayOf("antiphonFK"),
        onDelete = CASCADE,
        onUpdate = CASCADE
    ), ForeignKey(
        entity = LHThemeEntity::class,
        parentColumns = arrayOf("themeID"),
        childColumns = arrayOf("themeFK"),
        onDelete = SET_NULL,
        onUpdate = CASCADE
    ), ForeignKey(
        entity = EpigraphEntity::class,
        parentColumns = arrayOf("epigraphID"),
        childColumns = arrayOf("epigraphFK"),
        onDelete = SET_NULL,
        onUpdate = CASCADE
    )]
)
class LHPsalmodyEntity {
    @JvmField
    @ColumnInfo(name = "groupFK")
    var grupoFK = 0

    @JvmField
    @ColumnInfo(name = "readingFK", index = true)
    var salmoFK = 0

    @JvmField
    @ColumnInfo(name = "order")
    var orden = 0

    @JvmField
    @ColumnInfo(name = "antiphonFK", defaultValue = "0", index = true)
    var antifonaId: Int = 0

    @JvmField
    @ColumnInfo(name = "themeFK", defaultValue = "NULL", index = true)
    var temaFK: Int? = null

    @JvmField
    @ColumnInfo(name = "epigraphFK", defaultValue = "NULL", index = true)
    var epigrafeFK: Int? = null

    @JvmField
    @ColumnInfo(name = "part", defaultValue = "NULL")
    var parte: Int? = 0
    fun getParte(): String {
        return if (parte == null || parte == 0) "" else parte.toString()
    }
}